File: crates/agents-crew-cli/src/args.rs
use clap::{Args, Parser, Subcommand, ValueEnum};
use std::path::PathBuf;

#[derive(Debug, Parser)]
#[command(
    name = "crew",
    version,
    about = "Rust-enforced multi-agent crew orchestration"
)]
pub struct Cli {
    #[arg(long, global = true, default_value = ".")]
    pub workspace: PathBuf,
    #[arg(long, global = true)]
    pub json: bool,
    #[command(subcommand)]
    pub command: Command,
}

#[derive(Debug, Subcommand)]
pub enum Command {
    Init(InitArgs),
    Ui(UiArgs),
    Start(StartArgs),
    Run(GoalArgs),
    Plan(GoalArgs),
    Status(RunSelector),
    Resume(RunSelector),
    Pause(RunSelector),
    Approve(ApprovalArgs),
    Reject(ApprovalArgs),
    Cancel(RunSelector),
    Doctor,
    Template {
        #[command(subcommand)]
        command: TemplateCommand,
    },
    Config {
        #[command(subcommand)]
        command: ConfigCommand,
    },
    Plugin {
        #[command(subcommand)]
        command: PluginCommand,
    },
    Worker {
        #[command(subcommand)]
        command: WorkerCommand,
    },
    Manager {
        #[command(subcommand)]
        command: ManagerCommand,
    },
}

#[derive(Debug, Args)]
pub struct InitArgs {
    #[arg(long)]
    pub non_interactive: bool,
    #[arg(long)]
    pub force: bool,
}

#[derive(Debug, Args)]
pub struct UiArgs {
    #[arg(long, default_value_t = 0)]
    pub port: u16,
    #[arg(long)]
    pub no_open: bool,
}

#[derive(Debug, Args)]
pub struct StartArgs {
    pub template_id: String,
    #[arg(long, required = true)]
    pub goal: String,
    #[arg(long = "expectation")]
    pub expectations: Vec<String>,
    #[arg(long = "acceptance")]
    pub acceptance_criteria: Vec<String>,
    #[arg(long = "constraint")]
    pub constraints: Vec<String>,
}

#[derive(Debug, Args)]
pub struct GoalArgs {
    #[arg(required = true, num_args = 1..)]
    pub goal: Vec<String>,
}

#[derive(Debug, Args, Default)]
pub struct RunSelector {
    pub id: Option<String>,
    #[arg(long, conflicts_with = "id")]
    pub run: Option<String>,
}

impl RunSelector {
    #[must_use]
    pub fn selected(&self) -> Option<&str> {
        self.id.as_deref().or(self.run.as_deref())
    }
}

#[derive(Debug, Args)]
pub struct ApprovalArgs {
    pub approval_id: String,
    #[arg(long)]
    pub run: Option<String>,
}

#[derive(Debug, Clone, Copy, ValueEnum)]
pub enum TemplateScopeArg {
    Global,
    Workspace,
}

#[derive(Debug, Subcommand)]
pub enum TemplateCommand {
    List,
    Show { id: String },
    Validate { id: String },
    Delete {
        id: String,
        #[arg(long, value_enum)]
        scope: TemplateScopeArg,
    },
}

#[derive(Debug, Subcommand)]
pub enum ConfigCommand {
    Validate,
    Show,
}

#[derive(Debug, Subcommand)]
pub enum PluginCommand {
    List,
    Install {
        host: String,
        #[arg(long)]
        force: bool,
    },
    Doctor {
        host: String,
    },
    Uninstall {
        host: String,
    },
}

#[derive(Debug, Subcommand)]
pub enum WorkerCommand {
    Probe,
    Run { worker: String, task: PathBuf },
}

#[derive(Debug, Subcommand)]
pub enum ManagerCommand {
    Start {
        #[arg(long, required = true)]
        goal: String,
        #[arg(long, required = true)]
        host: String,
    },
    Step {
        #[arg(long)]
        run: String,
    },
    Submit {
        #[arg(long)]
        run: String,
        #[arg(long)]
        action: String,
        #[arg(long)]
        result: PathBuf,
    },
}

#[cfg(test)]
mod tests {
    use super::*;
    use clap::Parser;

    #[test]
    fn parses_template_start_with_durable_prompt_fields() {
        let cli = Cli::try_parse_from([
            "agents-crew",
            "start",
            "fullstack-review",
            "--goal",
            "ship feature",
            "--expectation",
            "keep compatibility",
            "--acceptance",
            "tests pass",
        ])
        .unwrap();
        let Command::Start(args) = cli.command else {
            panic!("expected start command");
        };
        assert_eq!(args.template_id, "fullstack-review");
        assert_eq!(args.expectations, ["keep compatibility"]);
        assert_eq!(args.acceptance_criteria, ["tests pass"]);
    }

    #[test]
    fn resume_accepts_positional_run_id() {
        let cli = Cli::try_parse_from(["agents-crew", "resume", "run-123"]).unwrap();
        let Command::Resume(selector) = cli.command else {
            panic!("expected resume command");
        };
        assert_eq!(selector.selected(), Some("run-123"));
    }

    #[test]
    fn parses_loopback_ui_options() {
        let cli = Cli::try_parse_from(["agents-crew", "ui", "--port", "4815", "--no-open"])
            .unwrap();
        let Command::Ui(args) = cli.command else {
            panic!("expected ui command");
        };
        assert_eq!(args.port, 4815);
        assert!(args.no_open);
    }
}

File: src/cli/args.ts
export interface ParsedArgs { workspace: string; json: boolean; command: string; args: Record<string, unknown> }

function takeValue(values: string[], index: number, flag: string): [string, number] {
  const value = values[index + 1];
  if (value === undefined || value.startsWith('--')) throw new Error(`${flag} requires a value`);
  return [value, index + 1];
}

export function parseArgs(argv: string[]): ParsedArgs {
  const values = [...argv];
  let workspace = '.';
  let json = false;
  const commandTokens: string[] = [];
  for (let index = 0; index < values.length; index += 1) {
    const value = values[index] as string;
    if (value === '--workspace') { [workspace, index] = takeValue(values, index, value); continue; }
    if (value === '--json') { json = true; continue; }
    commandTokens.push(value);
  }
  const command = commandTokens.shift();
  if (!command) throw new Error('missing command');
  const args: Record<string, unknown> = {};
  const positional: string[] = [];
  const repeated: Record<string, string[]> = { expectation: [], acceptance: [], constraint: [] };
  for (let index = 0; index < commandTokens.length; index += 1) {
    const value = commandTokens[index] as string;
    if (!value.startsWith('--')) { positional.push(value); continue; }
    const key = value.slice(2).replaceAll('-', '_');
    if (['non_interactive', 'force', 'no_open', 'yes', 'binary_only'].includes(key)) { args[key] = true; continue; }
    const [next, consumed] = takeValue(commandTokens, index, value);
    index = consumed;
    if (key === 'expectation' || key === 'acceptance' || key === 'constraint') repeated[key]?.push(next);
    else args[key] = next;
  }
  if (command === 'start') {
    args.template_id = positional[0];
    args.goal = args.goal ?? positional.slice(1).join(' ');
    args.expectations = repeated.expectation;
    args.acceptance_criteria = repeated.acceptance;
    args.constraints = repeated.constraint;
  } else if (['run', 'plan'].includes(command)) args.goal = positional.join(' ');
  else if (['status', 'resume', 'pause', 'cancel'].includes(command)) args.run_id = args.run ?? positional[0];
  else if (['approve', 'reject'].includes(command)) { args.approval_id = positional[0]; args.run_id = args.run; }
  else if (command === 'ui') args.port = Number(args.port ?? 0);
  else if (['template', 'config', 'plugin', 'worker', 'manager', 'agent'].includes(command)) {
    args.subcommand = positional[0];
    args.positional = positional.slice(1);
  } else args.positional = positional;
  return { workspace, json, command, args };
}
